File: apps/api/plane/app/serializers/custom_field.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.
#
# GAM addition: custom fields

from rest_framework import serializers

from plane.db.models import CustomField

from .base import BaseSerializer


class CustomFieldSerializer(BaseSerializer):
    class Meta:
        model = CustomField
        fields = ["id", "project_id", "name", "field_type", "options", "sequence"]
        read_only_fields = ["workspace", "project"]

    def validate_name(self, name):
        name = name.strip()
        if not name:
            raise serializers.ValidationError("The field needs a name.")
        return name

    def validate_options(self, options):
        if not isinstance(options, list):
            raise serializers.ValidationError("Options must be a list.")
        cleaned = []
        for option in options:
            text = str(option).strip()
            if text and text not in cleaned:
                cleaned.append(text)
        return cleaned

File: packages/types/src/custom-field.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 *
 * GAM addition: custom fields on work items.
 */

export type TCustomFieldType = "text" | "number" | "date" | "select" | "checkbox";

export interface ICustomField {
  id: string;
  project_id: string;
  name: string;
  field_type: TCustomFieldType;
  options: string[];
  sequence: number;
}

/** A work item's values by field id ("" = empty; checkboxes are "true" or "") */
export type TCustomFieldValues = Record<string, string>;
